From ddcab2adc5af0328837b7fa44d8c8af72836ff88 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:51:47 -0400
Subject: [fs] nfsd: hash deleg stateid only on successful nfs4_set_delegation

Message-id: <1403560314-32525-166-git-send-email-bfields@redhat.com>
Patchwork-id: 84599
O-Subject: [Patch RHEL7 165/172] nfsd4: hash deleg stateid only on successful nfs4_set_delegation
Bugzilla: 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Benny Halevy <bhalevy@primarydata.com>

We don't want the stateid to be found in the hash table before the delegation
is granted.

Currently this is protected by the client_mutex, but we want to break that
up and this is a necessary step toward that goal.

Signed-off-by: Benny Halevy <bhalevy@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 3fb87d13ce29637e0ae005fa66a4f917cce6fede
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 63e7645..e5197d9 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -375,7 +375,6 @@ alloc_init_deleg(struct nfs4_client *clp, struct nfs4_ol_stateid *stp, struct sv
  dp = delegstateid(nfs4_alloc_stid(clp, deleg_slab));
  if (dp == NULL)
   return dp;
- dp->dl_stid.sc_type = NFS4_DELEG_STID;
  /*
   * delegation seqid's are never incremented.  The 4.1 special
   * meaning of seqid 0 isn't meaningful, really, but let's avoid
@@ -438,6 +437,7 @@ hash_delegation_locked(struct nfs4_delegation *dp, struct nfs4_file *fp)
 {
  lockdep_assert_held(&state_lock);
 
+ dp->dl_stid.sc_type = NFS4_DELEG_STID;
  list_add(&dp->dl_perfile, &fp->fi_delegations);
  list_add(&dp->dl_perclnt, &dp->dl_stid.sc_client->cl_delegations);
 }
-- 
1.7.1